\documentclass[a4paper]{book}
\usepackage[times,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Title]\AsIs{ScRNA-seq workflow CONCLUS - from CONsensus CLUSters to a meaningful CONCLUSion.}
\item[Version]\AsIs{0.0.0.9000}
\item[Authors]\AsIs{Polina Pavlovich, Konstantin Chukreev, Christophe Lancrin}
\item[Maintainer]\AsIs{Polina Pavlovich }\email{pavlovich@phystech.edu}\AsIs{}
\item[Description]\AsIs{CONCLUS is a tool for robust clustering, positive marker features selection
and collecting info about marker genes using web scraping from MGI, UniProt and NCBI
websites. The tool was developed to distinguish not only cell types from different tissues
but also characterize heterogeneity within a tissue or characterize stages of a differentiation process.}
\item[Depends]\AsIs{R (>= 3.4)}
\item[License]\AsIs{GPL-3}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[RoxygenNote]\AsIs{6.1.1.9000}
\item[Imports]\AsIs{BiocParallel (>= 1.14.2), scran (>= 1.8.4), ggplot2 (>= 3.0.0),
scater (>= 1.8.4), Matrix (>= 1.2-14), monocle (>= 2.8.0), SingleCellExperiment (>= 1.2.0),
dbscan (>= 1.1-2), pheatmap (>= 1.0.10), fpc (>= 2.1-11.1), dynamicTreeCut (>= 1.63-1),
factoextra (>= 1.0.5), digest (>= 0.6.15), KEGGREST (>= 1.20.2), RColorBrewer (>= 1.1-2),
doParallel (>= 1.0.14), matrixStats (>= 0.54.0), AnnotationDbi (>= 1.42.1), dplyr (>= 0.7.6),
biomaRt (>= 2.36.1), org.Mm.eg.db (>= 3.6.0), xlsx (>= 0.6.1), DataCombine (>= 0.2.21),
grDevices, S4Vectors (>= 0.18.3), Biobase (>= 2.40.0), foreach (>= 1.4.4)}
\item[Suggests]\AsIs{knitr, rmarkdown, prettydoc}
\item[VignetteBuilder]\AsIs{knitr}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{addClusteringManually}{addClusteringManually}{addClusteringManually}
%
\begin{Description}\relax
The function replaces the content of the column "clusters" in the colData(sceObject) 
with the clustering provided in the user table.
The function will return the sceObject with cells which intersect with the cells from the input table.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
addClusteringManually(fileName, sceObject, dataDirectory, experimentName,
  columnName = "clusters")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{fileName}] a file with the clustering solution (for example, from previous CONCLUS runs).

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which appears in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{columnName}] name of the column with the clusters.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A SingleCellExperiment object with the created/renewed column "clusters" in the colData(sceObject).
\end{Value}
\inputencoding{utf8}
\HeaderA{calculateClustersSimilarity}{Having cells similarity, calculate clusters similarity.}{calculateClustersSimilarity}
%
\begin{Description}\relax
Having cells similarity, calculate clusters similarity.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
calculateClustersSimilarity(cellsSimilarityMatrix, sceObject,
  clusteringMethod)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{cellsSimilarityMatrix}] a similarity matrix, one of the results of conclus::clusterCellsInternal() function.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list contating the cluster similarity matrix and cluster names (order).
\end{Value}
\inputencoding{utf8}
\HeaderA{choosePalette}{Choose palette for a plot.}{choosePalette}
%
\begin{Description}\relax
It is an internal function usually applied for choosing the palette for clusters.
Depending if the number of clusters is more than 12 or not, one of two built-in palettes will be applied.
If you give your vector of colors, the function will not change them.
If the number of clusters is more than 26, it will copy colors to get the needed length of the palette.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
choosePalette(colorPalette, clustersNumber)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{colorPalette}] Either "default" or a vector of colors, for example c("yellow", "\#CC79A7").

\item[\code{clustersNumber}] number of clusters in the output palette.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Color palette with the number of colors equal to the clusterNumber parameter.
\end{Value}
\inputencoding{utf8}
\HeaderA{clusterCellsInternal}{Cluster cells and get similarity matrix of cells.}{clusterCellsInternal}
%
\begin{Description}\relax
The function returns consensus clusters by using hierarchical clustering on the similarity matrix of cells.
It provides two options: to specify an exact number of clusters (with clusterNumber parameter)
or to select the depth of splitting (deepSplit parameter).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
clusterCellsInternal(dbscanMatrix, sceObject, clusterNumber = 0,
  deepSplit, cores = 14, clusteringMethod = "ward.D2")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dbscanMatrix}] an output matrix of conclus::runDBSCAN() function.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{clusterNumber}] a parameter, specifying the exact number of cluster.

\item[\code{deepSplit}] a parameter, specifying how deep we will split the clustering tree. It takes integers from 1 to 4.

\item[\code{cores}] maximum number of jobs that CONCLUS can run in parallel.

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A SingleCellExperiment object with modified/created "clusters" column in the colData, and cells similarity matrix.
\end{Value}
\inputencoding{utf8}
\HeaderA{exportClusteringResults}{exportClusteringResults}{exportClusteringResults}
%
\begin{Description}\relax
The function saves clustering results into a table. Row names are cell names in the same order as in the sceObject.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
exportClusteringResults(sceObject, dataDirectory, experimentName, fileName)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which appears at the beginning of the file name 
(supposed to be the same for one experiment during the workflow).

\item[\code{fileName}] the rest of output file name.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{exportMatrix}{Export matrix to a file.}{exportMatrix}
%
\begin{Description}\relax
The function allows you to export a matrix to a .csv file with a hard-coded filename (according to experimentName) 
in the "dataDirectory/output\_tables" directory for further analysis.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
exportMatrix(matrix, dataDirectory, experimentName, name)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{matrix}] your matrix (e.g., expression matrix)

\item[\code{dataDirectory}] CONCLUS output directory for a given experiment (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear at the beginning of the filenames 
(supposed to be the same for one experiment during the workflow).

\item[\code{name}] name of the file. Will be placed after the experimentName header.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{generateTSNECoordinates}{Generate and save t-SNE coordinates with selected parameters.}{generateTSNECoordinates}
%
\begin{Description}\relax
The function generates several t-SNE coordinates based on given perplexity and ranges of PCs. 
Final number of t-SNE plots is length(PCs)*length(perplexities)
It writes coordinates in "dataDirectory/tsnes" subfolder.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
generateTSNECoordinates(sceObject, dataDirectory, experimentName,
  randomSeed = 42, cores = 14, PCs = c(4, 6, 8, 10, 20, 40, 50),
  perplexities = c(30, 40))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{randomSeed}] random seed for reproducibility.

\item[\code{cores}] maximum number of jobs that CONCLUS can run in parallel.

\item[\code{PCs}] a vector of first principal components.
For example, to take ranges 1:5 and 1:10 write c(5, 10).

\item[\code{perplexities}] a vector of perplexity (t-SNE parameter).
\end{ldescription}
\end{Arguments}
%
\begin{Value}
An object with t-SNE results (coordinates for each plot).
\end{Value}
\inputencoding{utf8}
\HeaderA{getGenesInfo}{Collect genes information to one table.}{getGenesInfo}
%
\begin{Description}\relax
The function takes a data frame containing gene symbols and (or) ENSEMBL IDs and returns
a data frame with such information as gene name, feature type, chromosome,
gene IDs in different annotations, knockout information from MGI, a summary from NCBI 
and UniProt, and whether or not a gene belongs to GO terms containing proteins on the cell surface or 
involved in secretion.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getGenesInfo(genes, databaseDir = system.file("extdata", package =
  "conclus"), groupBy = "clusters", orderGenes = "initial",
  getUniprot = TRUE, silent = FALSE, coresGenes = 20)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{genes}] a data frame with the first column called "geneName" containing gene symbols and (or) ENSEMBL IDs.
Other columns are optional. For example, the second column could be "clusters" with the name of the cluster 
for which the gene is a marker.

\item[\code{databaseDir}] a path to the database provided with CONCLUS called "Mmus\_gene\_database\_secretedMol.tsv".

\item[\code{groupBy}] a column in the input table used for grouping the genes in the output tables.
This option is useful if a table contains genes from different clusters.

\item[\code{orderGenes}] if "initial" then the order of genes will not be changed.

\item[\code{getUniprot}] boolean, whether to get information from UniProt or not. Default is TRUE.
Sometimes, the connection to the website is not reliable. 
If you tried a couple of times and it failed, select FALSE.

\item[\code{silent}] whether to show messages from intermediate steps or not.

\item[\code{coresGenes}] maximum number of jobs that the function can run in parallel.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Returns a data frame.
\end{Value}
\inputencoding{utf8}
\HeaderA{getMarkerGenes}{Get top N marker genes from each cluster.}{getMarkerGenes}
%
\begin{Description}\relax
This function reads results of conclus::rankGenes() from "dataDirectory/marker\_genes" and selects top N markers for each cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getMarkerGenes(dataDirectory, sceObject, genesNumber = 14,
  experimentName, removeDuplicates = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dataDirectory}] output directory for a run of CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{genesNumber}] top N number of genes to get from one cluster.

\item[\code{experimentName}] name of the experiment which appears in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{removeDuplicates}] boolean, if duplicated genes must be deleted or not.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data frame where the first columns are marker genes ("geneName") and 
the second column is the groups ("clusters").
\end{Value}
\inputencoding{utf8}
\HeaderA{initialisePath}{Create all needed directories for CONCLUS output.}{initialisePath}
%
\begin{Description}\relax
Create all needed directories for CONCLUS output.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
initialisePath(dataDirectory)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dataDirectory}] output directory for a given CONCLUS run (supposed to be the same for one experiment during the workflow).
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{normaliseCountMatrix}{normaliseCountMatrix}{normaliseCountMatrix}
%
\begin{Description}\relax
Create a SingleCellExperiment object and perform normalization. The same as conclus::normalizeCountMatrix.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
normaliseCountMatrix(countMatrix, species, method = "default",
  sizes = c(20, 40, 60, 80, 100), rowData = NULL, colData = NULL,
  alreadyCellFiltered = FALSE, runQuickCluster = TRUE,
  databaseDir = system.file("extdata", package = "conclus"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{countMatrix}] a matrix with non-normalised gene expression.

\item[\code{species}] either 'mmu' or 'human'.

\item[\code{method}] a method of clustering: available option is "default" using scran and scater.

\item[\code{sizes}] a vector of size factors from scran::computeSumFactors() function.

\item[\code{rowData}] a data frame with information about genes

\item[\code{colData}] a data frame with information about cells

\item[\code{alreadyCellFiltered}] if TRUE, cells quality check and filtering will not be applied. 
However, the function may delete some cells if they have negative size factors after scran::computeSumFactors.

\item[\code{runQuickCluster}] if scran::quickCluster() function must be applied.
Usually, it allows to improve normalization for medium-size count matrices. 
However, it is not recommended for datasets with less than 200 cells and
may take too long for datasets with more than 10000 cells.

\item[\code{databaseDir}] a path to annotation database provided with CONCLUS called 
"Mmus\_gene\_database\_secretedMol.tsv" (only for MusMusculus 'mmu').
The function will work also without the database but slower because it will retrieve genes info from biomaRt.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A SingleCellExperiment object with normalized gene expression, colData, and rowData.
\end{Value}
\inputencoding{utf8}
\HeaderA{plotCellHeatmap}{Save markers heatmap.}{plotCellHeatmap}
%
\begin{Description}\relax
This function plots heatmap with marker genes on rows and clustered cells on columns.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCellHeatmap(markersClusters, sceObject, dataDirectory, experimentName,
  fileName, meanCentered = TRUE, colorPalette = "default",
  statePalette = "default", clusteringMethod = "ward.D2",
  orderClusters = FALSE, orderGenes = FALSE, returnPlot = FALSE,
  saveHeatmapTable = FALSE, width = 10, height = 8.5, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{markersClusters}] a data frame where the first column is "geneName" containing genes names from sceObject, 
and the second column is corresponding "clusters". All names from that column must come from the column "clusters" in the colData(sceObject).
The data frame can be obtained from conclus::getMarkerGenes() function or created manually.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory of a given CONCLUS run (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which appears in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{fileName}] name of the ouput file

\item[\code{meanCentered}] boolean, should mean centering be applied to the expression data or not.

\item[\code{colorPalette}] "default" or a vector of colors for the column "clusters" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{statePalette}] "default" or a vector of colors for the column "state" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.

\item[\code{orderClusters}] boolean, should the heatmap be structured by clusters.

\item[\code{orderGenes}] boolean, should the heatmap be structured by genes.

\item[\code{returnPlot}] boolean, whether to return a ggplot object with the plot or not.

\item[\code{saveHeatmapTable}] boolean, whether to save the expression matrix used for heatmap into a .csv file or not.
The file will be saved into 'dataDirectory/output\_tables' with the same name as the .pdf plot.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other parameters from pdf() and pheatmap() functions.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object of the plot if needed. The function saves pdf in "dataDirectiry/pictures" folder.
\end{Value}
\inputencoding{utf8}
\HeaderA{plotCellSimilarity}{Save a cells similarity matrix.}{plotCellSimilarity}
%
\begin{Description}\relax
This function plots similarity matrix as a heatmap, so one can see similarity between parts of different clusters.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotCellSimilarity(sceObject, cellsSimilarityMatrix, dataDirectory,
  experimentName, colorPalette = "default", statePalette = "default",
  clusteringMethod = "ward.D2", orderClusters = FALSE,
  plotPDF = TRUE, returnPlot = FALSE, width = 7, height = 6, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{cellsSimilarityMatrix}] an output matrix from the conclus::clusterCellsInternal() function.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{colorPalette}] "default" or a vector of colors for the column "clusters" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{statePalette}] "default" or a vector of colors for the column "state" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.

\item[\code{orderClusters}] boolean, order clusters or not.

\item[\code{plotPDF}] if TRUE export to pdf, if FALSE export to png. 
FALSE is recommended for datasets with more than 2500 cells due to large pdf file size.

\item[\code{returnPlot}] boolean, return plot or not. Default if FALSE.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other parameters of pdf(), pheatmap() and png() functions.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object or nothing (depends on the returnPlot parameter).
It saves the pdf in "dataDirectory/pictures" folder.
\end{Value}
\inputencoding{utf8}
\HeaderA{plotClusteredTSNE}{Plot t-SNE. Addtionally, it can highlight clusters or states.}{plotClusteredTSNE}
%
\begin{Description}\relax
Plot t-SNE. Addtionally, it can highlight clusters or states.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotClusteredTSNE(sceObject, dataDirectory, experimentName,
  tSNEresExp = "", colorPalette = "default", PCs = c(4, 6, 8, 10, 20,
  40, 50), perplexities = c(30, 40), columnName = "clusters",
  returnPlot = FALSE, width = 6, height = 5, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{tSNEresExp}] if t-SNE coordinates were generated in a different CONCLUS run, you can use them without renaming the files.
Please copy tsnes folder from the source run to the current one and write that experimentName in the tSNEresExp argument.

\item[\code{colorPalette}] "default" or a vector of colors for the column "clusters" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{PCs}] vector of PCs (will be specified in filenames).

\item[\code{perplexities}] vector of perplexities (will be specified in filenames).

\item[\code{columnName}] name of the column to plot on t-SNE dimensions.

\item[\code{returnPlot}] boolean, return plot or not.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other arguments of the pdf() function.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object or nothing (depends on the returnPlot parameter).
\end{Value}
\inputencoding{utf8}
\HeaderA{plotClustersSimilarity}{Save a similarity cluster matrix.}{plotClustersSimilarity}
%
\begin{Description}\relax
Save a similarity cluster matrix.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotClustersSimilarity(clustersSimilarityMatrix, sceObject, dataDirectory,
  experimentName, colorPalette, statePalette, clusteringMethod,
  returnPlot = FALSE, width = 7, height = 5.5, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{clustersSimilarityMatrix}] a matrix, result of conclus::calculateClustersSimilarity() function.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{colorPalette}] "default" or a vector of colors for the column "clusters" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{statePalette}] "default" or a vector of colors for the column "state" in the colData, for example c("yellow", "\#CC79A7").

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.

\item[\code{returnPlot}] boolean, return plot or not.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other parameters of pdf() and pheatmap() functions.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object or nothing (depends on returnPlot parameter). It saves the pdf in "dataDirectory/pictures" folder.
\end{Value}
\inputencoding{utf8}
\HeaderA{plotGeneExpression}{plotGeneExpression}{plotGeneExpression}
%
\begin{Description}\relax
The function saves a t-SNE plot colored by expression of a given gene. 
Warning: filename with t-SNE results is hardcoded, so please don't rename the output file.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotGeneExpression(geneName, experimentName, dataDirectory,
  graphsDirectory = "pictures", sceObject, tSNEpicture = 1,
  commentName = "", palette = c("grey", "red", "#7a0f09", "black"),
  returnPlot = FALSE, savePlot = TRUE, alpha = 1, limits = NA,
  pointSize = 1, width = 6, height = 5, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{geneName}] name of the gene you want to plot.

\item[\code{experimentName}] name of the experiment which appears in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{graphsDirectory}] name of the subdirectory where to put graphs. Default is "dataDirectory/pictures".

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{tSNEpicture}] number of the picture you want to use for plotting. 
Please check "dataDirectory/tsnes" or "dataDirectory/pictures/tSNE\_pictures/clusters" to get the number, it is usually from 1 to 14.

\item[\code{commentName}] comment you want to specify in the filename.

\item[\code{palette}] color palette for the legend.

\item[\code{returnPlot}] boolean, should the function return a ggplot object or not.

\item[\code{savePlot}] boolean, should the function export the plot to pdf or not.

\item[\code{alpha}] opacity of the points of the plot.

\item[\code{limits}] range of the gene expression shown in the legend.
This option allows generating t-SNE plots with equal color
scale to compare the expression of different genes. By default, limits are the range
of expression of a selected gene.

\item[\code{pointSize}] size of the point.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other parameters of the pdf() function.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object of the plot if needed.
\end{Value}
\inputencoding{utf8}
\HeaderA{rankGenes}{Rank marker genes by statistical significance.}{rankGenes}
%
\begin{Description}\relax
This function searches marker genes for each cluster. It saves tables in the "dataDirectory/marker\_genes" directory,
one table per cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
rankGenes(sceObject, clustersSimilarityMatrix, dataDirectory,
  experimentName, column = "clusters")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{clustersSimilarityMatrix}] matrix, result of conclus::calculateClustersSimilarity() function.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{column}] name of the column with a clustering result.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{runClustering}{DBSCAN clustering on t-SNE results.}{runClustering}
%
\begin{Description}\relax
This function provides consensus DBSCAN clustering based on the results of t-SNE. 
You can tune algorithm parameters in options to get the number of clusters you want.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runClustering(tSNEResults, sceObject, dataDirectory, experimentName,
  epsilon = c(1.3, 1.4, 1.5), minPoints = c(3, 4), k = 0,
  deepSplit = 4, clusteringMethod = "ward.D2", cores = 14,
  deleteOutliers = TRUE, PCs = c(4, 6, 8, 10, 20, 40, 50),
  perplexities = c(30, 40), randomSeed = 42)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{tSNEResults}] the result of conclus::generateTSNECoordinates() function.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory of a given CONCLUS run (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which appears in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{epsilon}] a parameter of fpc::dbscan() function.

\item[\code{minPoints}] a parameter of fpc::dbscan() function.

\item[\code{k}] preferred number of clusters. Alternative to deepSplit.

\item[\code{deepSplit}] intuitive level of clustering depth. Options are 1, 2, 3, 4.

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.

\item[\code{cores}] maximum number of jobs that CONCLUS can run in parallel.

\item[\code{deleteOutliers}] Whether cells which were often defined as outliers by dbscan must be deleted.
It will require recalculating of the similarity matrix of cells. Default is FALSE.
Usually those cells appear in an "outlier" cluster and can be easier distinguished and deleted later
if necessary.

\item[\code{PCs}] a vector of first principal components.
For example, to take ranges 1:5 and 1:10 write c(5, 10).

\item[\code{perplexities}] a vector of perplexity for t-SNE.

\item[\code{randomSeed}] random seed for reproducibility.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list containing filtered from outliers SingleCellExperiment object and cells similarity matrix.
\end{Value}
\inputencoding{utf8}
\HeaderA{runCONCLUS}{Run CONCLUS in one click}{runCONCLUS}
\keyword{CONCLUS}{runCONCLUS}
%
\begin{Description}\relax
This function performs core CONCLUS workflow. It generates PCA and t-SNE coordinates, 
runs DBSCAN, calculates similarity matrices of cells and clusters, assigns cells to clusters,
searches for positive markers for each cluster. The function saves plots and tables into dataDirectory.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runCONCLUS(sceObject, dataDirectory, experimentName,
  colorPalette = "default", statePalette = "default",
  clusteringMethod = "ward.D2", epsilon = c(1.3, 1.4, 1.5),
  minPoints = c(3, 4), k = 0, PCs = c(4, 6, 8, 10, 20, 40, 50),
  perplexities = c(30, 40), randomSeed = 42, deepSplit = 4,
  preClustered = F, orderClusters = FALSE, cores = 14,
  plotPDFcellSim = TRUE, deleteOutliers = TRUE,
  tSNEalreadyGenerated = FALSE, tSNEresExp = "")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your data.

\item[\code{dataDirectory}] CONCLUS will create this directory if it doesn't exist and store there all output files.

\item[\code{experimentName}] most of output file names of CONCLUS are hardcoded.
experimentName will stay at the beginning of each output file name to
distinguish different runs easily.

\item[\code{colorPalette}] a vector of colors for clusters.

\item[\code{statePalette}] a vector of colors for states.

\item[\code{clusteringMethod}] a clustering methods passed to hclust() function.

\item[\code{epsilon}] a parameter of fpc::dbscan() function.

\item[\code{minPoints}] a parameter of fpc::dbscan() function.

\item[\code{k}] preferred number of clusters. Alternative to deepSplit. A parameter of cutree() function.

\item[\code{PCs}] a vector of first principal components.
For example, to take ranges 1:5 and 1:10 write c(5, 10).

\item[\code{perplexities}] a vector of perplexity for t-SNE.

\item[\code{randomSeed}] random seed for reproducibility.

\item[\code{deepSplit}] intuitive level of clustering depth. Options are 1, 2, 3, 4.

\item[\code{preClustered}] if TRUE, it will not change the column clusters after the run.
However, it will anyway run DBSCAN to calculate similarity matrices.

\item[\code{orderClusters}] can be either FALSE (default) of "name".
If "name", clusters in the similarity matrix of cells will be ordered by name.

\item[\code{cores}] maximum number of jobs that CONCLUS can run in parallel.

\item[\code{plotPDFcellSim}] if FALSE, the similarity matrix of cells will be saved in png format.
FALSE is recommended for count matrices with more than 2500 cells due to large pdf file size.

\item[\code{deleteOutliers}] whether cells which were often defined as outliers by dbscan must be deleted.
It will require recalculating of the similarity matrix of cells. Default is FALSE.
Usually those cells form a separate "outlier" cluster and can be easier distinguished and deleted later
if necessary.

\item[\code{tSNEalreadyGenerated}] if you already ran CONCLUS ones and have t-SNE coordinated saved
You can set TRUE to run the function faster since it will skip the generation of t-SNE coordinates and use the stored ones. 
Option TRUE requires t-SNE coordinates to be located in your 'dataDirectory/tsnes' directory.

\item[\code{tSNEresExp}] experimentName of t-SNE coordinates which you want to use.
This argument allows copying and pasting t-SNE coordinates between different CONCLUS runs without renaming the files.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A SingleCellExperiment object.
\end{Value}
\inputencoding{utf8}
\HeaderA{runDBSCAN}{Run clustering iterations with selected parameters using DBSCAN.}{runDBSCAN}
%
\begin{Description}\relax
This function returns a matrix of clustering iterations of DBSCAN.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
runDBSCAN(tSNEResults, sceObject, dataDirectory, experimentName,
  cores = 14, epsilon = c(1.3, 1.4, 1.5), minPoints = c(3, 4))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{tSNEResults}] results of conclus::generateTSNECoordinates() function.

\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory for CONCLUS (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames 
(supposed to be the same for one experiment during the workflow).

\item[\code{cores}] maximum number of jobs that CONCLUS can run in parallel.

\item[\code{epsilon}] a fpc::dbscan() parameter.

\item[\code{minPoints}] a fpc::dbscan() parameter.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A matrix of DBSCAN results.
\end{Value}
\inputencoding{utf8}
\HeaderA{saveGenesInfo}{Save gene information into a table or tables for multiple inputs.}{saveGenesInfo}
%
\begin{Description}\relax
This function runs conclus::getGenesInfo() function for all tables into the inputDir 
and saves the result into the outputDir.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
saveGenesInfo(dataDirectory = "", inputDir = "", outputDir = "",
  pattern = "", databaseDir = system.file("extdata", package =
  "conclus"), sep = ";", header = TRUE, startFromFile = 1,
  groupBy = "clusters", orderGenes = "initial", getUniprot = TRUE,
  silent = FALSE, coresGenes = 20)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{dataDirectory}] a directory with CONCLUS output. You can specify either 
dataDirectory, then inputDir and outputDir will be hardcoded, or inputDir and outputDir only.
The first is recommended during running CONCLUS workflow when the second option
is comfortable when you created input tables with genes manually.

\item[\code{inputDir}] input directory containing text files. These files can be obtained by 
applying conclus::saveMarkersLists() function or created manually. Each file must be a 
data frame with the first column called "geneName" containing gene symbols and (or) ENSEMBL IDs.

\item[\code{outputDir}] output directory.

\item[\code{pattern}] a pattern of file names to take.

\item[\code{databaseDir}] a path to the database "Mmus\_gene\_database\_secretedMol.tsv". It is provided with the conclus package.

\item[\code{sep}] a parameter of read.delim() function.

\item[\code{header}] whether or not your input files have a header.

\item[\code{startFromFile}] number of the input file to start with. The function approaches files one by one.
It uses web scraping method to collect publicly available info from MGI, NCBI and UniProt websites.
Sometimes, if the Internet connection is not reliable, the function can drop. 
In this case, it is comfortable to start from the failed file and not to redo the previous ones.

\item[\code{groupBy}] a column in the input table used for grouping the genes in the output tables.

\item[\code{orderGenes}] if "initial" then the order of genes will not be changed.

\item[\code{getUniprot}] boolean, whether to get information from UniProt or not. Default is TRUE.
Sometimes, the connection to the website is not reliable. 
If you tried a couple of times and it failed, select FALSE.

\item[\code{silent}] whether to show messages from intermediate steps or not.

\item[\code{coresGenes}] maximum number of jobs that the function can run in parallel.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
It saves text files either in the 'dataDirectory/marker\_genes/saveGenesInfo' or outputDir 
depending on whether you specify dataDirectory or (inpitDir and outputDir) explicitly.
\end{Value}
\inputencoding{utf8}
\HeaderA{saveMarkersLists}{Save top N marker genes for each cluster into a format suitable for conclus::saveGenesInfo() function.}{saveMarkersLists}
%
\begin{Description}\relax
The function takes the output files of conclus::rankGenes(), extracts top N markers and saves
them into the first "geneName" column of the output table. The second column "clusters" contains the 
name of the corresponding cluster.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
saveMarkersLists(experimentName, dataDirectory,
  inputDir = file.path(dataDirectory, "marker_genes"),
  outputDir = file.path(dataDirectory,
  paste0("marker_genes/markers_lists")), pattern = "genes.tsv",
  Ntop = 100)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{experimentName}] name of the experiment which appears at the beginning of the file name 
(supposed to be the same for one experiment during the workflow).

\item[\code{dataDirectory}] experiment directory (supposed to be the same for one experiment during the workflow).

\item[\code{inputDir}] input directory, usually "marker\_genes" created automatically after conclus::runCONCLUS().

\item[\code{outputDir}] output directory.

\item[\code{pattern}] a pattern of the input file names to take.

\item[\code{Ntop}] number of top markers to take from each cluster.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
It saves files into the outputDir. The number of files is equal to the number of clusters.
\end{Value}
\inputencoding{utf8}
\HeaderA{testClustering}{To check one iteration of clustering before running full workflow CONCLUS.}{testClustering}
%
\begin{Description}\relax
This function generates a single clustering iteration of CONCLUS to check whether
chosen parameters for dbscan are suitable for your data.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
testClustering(sceObject, dataDirectory, experimentName,
  dbscanEpsilon = 1.4, minPts = 5, perplexities = c(30),
  PCs = c(4), randomSeed = 42, width = 7, height = 7, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sceObject}] a SingleCellExperiment object with your experiment.

\item[\code{dataDirectory}] output directory (supposed to be the same for one experiment during the workflow).

\item[\code{experimentName}] name of the experiment which will appear in filenames (supposed to be the same for one experiment during the workflow).

\item[\code{dbscanEpsilon}] a parameter of fpc::dbscan() function.

\item[\code{minPts}] a parameter of fpc::dbscan() function.

\item[\code{perplexities}] vector of perplexities (t-SNE parameter).

\item[\code{PCs}] a vector of PCs for plotting.

\item[\code{randomSeed}] random seed for reproducibility.

\item[\code{width}] plot width.

\item[\code{height}] plot height.

\item[\code{...}] other pdf() arguments.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
t-SNE results, a distance graph plot, a t-SNE plot colored by test clustering solution.
\end{Value}
\printindex{}
\end{document}
